### ISSUE TITLE
replace LinkText with RichText?

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/joist#430]
In https://github.com/phetsims/scenery-phet/issues/316#issuecomment-313741189 (RichText link support), @jonathanolson said:

> • Didn't use LinkText, since it was cleaner overall to implement in RichText directly (can have formatting inside a link, which is potentially important).

@jonathanolson Since LinkText is used only in joist and molecule-polarity, should we replace it with RichText, then delete LinkText?

**Comment by jonathanolson** [comment]
> should we replace it with RichText, then delete LinkText?

I hadn't immediately handled a11y in RichText (isn't trivial), so no not yet. I should be able to implement it however (will tag this issue), so we can look into just using RichText.

Plan would be to determine the "non-marked-up" version of the linked text (decoding HTML entities, ignoring tags) for the accessibleLabel, and then do everything else LinkText does. For example:

```js
new RichText( 'A <a href="{{url}}">link has <b>markup</b></a>.', ... );
```
would presumably set accessibleLabel to `'link has markup'` (tags removed).

@jessegreenberg, does this sound like an appropriate approach? Would subscripts/superscripts need any additional changes with how they would be assigned to accessibleLabel?

**Comment by pixelzoom** [comment]
Any objection to changing from LinkText to RichText in `molecule-polarity.UnderDevelopmentPlane`?  This is the dialog that informs the user that "Real Molecules" screen is not implemented, and directs them (via a link) to the legacy Java version.  This dialog is temporary, and I'm not aware of any a11y requirements for it.

**Comment by jonathanolson** [comment]
> Any objection to changing from LinkText to RichText in molecule-polarity.UnderDevelopmentPlane?

No objections from me.

**Comment by jonathanolson** [comment]
I applied that approach to RichText.

Selecting it works well (using ?accessibility):
![richtext-a11y](https://user-images.githubusercontent.com/480878/27972835-f4dd020c-6314-11e7-8266-a9a883093e86.PNG)

but space and enter don't seem to trigger the link.

@jessegreenberg, can you see if I'm doing something wrong here? Once it's fixed, I believe we can move forward with replacing LinkText with RichText.

**Comment by jessegreenberg** [comment]
My only change was to add the href and target attributes to the created node (just like the tagName and label).  The links can now be triggered with "enter". "spacebar" typically doesn't trigger links.

The link ignored for security is still getting keyboard focus, ill take a look at that.

**Comment by jessegreenberg** [comment]
Never mind, its a checkbox that is behind the links in another demo that is invisible but still focusable.

**Comment by jessegreenberg** [comment]
Regarding https://github.com/phetsims/joist/issues/430#issuecomment-313760585m ideally the markup would be included in the accessibleLabel, screen readers sometimes pause or change inflection depending on how the text is marked up. @jonathanolson do we have to strip it out?

**Comment by jessegreenberg** [comment]
> Any objection to changing from LinkText to RichText in molecule-polarity.UnderDevelopmentPlane

None from me, I don't know of any a11y requirements for molecule-polarity either.

**Comment by jonathanolson** [comment]
> ideally the markup would be included in the accessibleLabel, screen readers sometimes pause or change inflection depending on how the text is marked up. @jonathanolson do we have to strip it out?

It sounds good to include the tags, however there are currently encoding and formatted-tag-list issues with doing so. See https://github.com/phetsims/scenery/issues/643.

If those issues can be resolved, I think it sounds absolutely great to include the markup in the accessibleLabel.

**Comment by pixelzoom** [comment]
LinkText has been replaced with RichText in molecule-polarity, see https://github.com/phetsims/molecule-polarity/issues/34.  The remaining uses are in joist (AboutDialog, UpdateNodes).

**Comment by jonathanolson** [comment]
Added including certain tags for accessibleLabel, so:
```js
new RichText( 'Supports <a href="{{phetWebsite}}"><em>links</em> with <b>markup</b></a>.', {
  links: {
    phetWebsite: 'https://phet.colorado.edu'
  }
} )
```
ends up with `"<em>‪links‬</em>‪ with ‬<b>‪markup‬</b>"` as the accessibleLabel.

**Comment by jessegreenberg** [comment]
This requires https://github.com/phetsims/scenery/issues/643 to be completed first.

**Comment by jessegreenberg** [comment]
rootNode's links look like this in the accessible DOM:

![screen shot 2017-08-10 at 4 49 58 pm](https://user-images.githubusercontent.com/6396244/29195680-f998b1e0-7dec-11e7-86fa-a62bb70c42e9.png)

And they sound great with VoiceOver.

**Comment by jessegreenberg** [comment]
I would like to work on making RichText more generally accessible in #662, but is there anything else to work on in this issue before deprecating LinkText?

**Comment by jonathanolson** [comment]
I can't think of anything else personally. I'll look into the replacement.

**Comment by pixelzoom** [comment]
>  anything else to work on in this issue before deprecating LinkText?

Sounds like we'll be able to delete (vs deprecate) LinkText after the replacement, correct?

**Comment by jonathanolson** [comment]
> Sounds like we'll be able to delete (vs deprecate) LinkText after the replacement, correct?

Correct.

**Comment by pixelzoom** [comment]
I converted AboutDialog and UpdateNodes to use RichText, then deleted LinkText.

@jonathanolson please review with relatively high priority.

**Comment by pixelzoom** [comment]
@jessegreenberg FYI, in case there are any a11y issues.

**Comment by pixelzoom** [comment]
@jonathanolson @jessegreenberg please note the high priority of this issue.

**Comment by jessegreenberg** [comment]
Thanks @pixelzoom. I reviewed the four commits between https://github.com/phetsims/joist/issues/430#issuecomment-322622394 and https://github.com/phetsims/joist/issues/430#issuecomment-376232459.

I tested links in UpdateDialog and AboutDialog and they are behaving as they should for a11y with `RichText`.

**Comment by pixelzoom** [comment]
This is top priority because it involved significant changes to joist, was assigned for review and labeled "high priority" 16 days ago.

**Comment by jonathanolson** [comment]
Looks good, sorry about the delay.
